lib/bolt/puppetfile.rb

```diff
@@ -1,27 +1,32 @@
 # frozen_string_literal: true
 
+require 'puppetfile-resolver'
+
 require 'bolt/error'
-require 'bolt/puppetfile/module'
+require 'bolt/puppetfile/forge_module'
+require 'bolt/puppetfile/forge_spec'
+require 'bolt/puppetfile/git_module'
+require 'bolt/puppetfile/git_spec'
 
 # This class manages the logical contents of a Puppetfile. It includes methods
 # for parsing a Puppetfile and its modules, resolving module dependencies,
 # and writing a Puppetfile.
 #
 module Bolt
   class Puppetfile
-    attr_reader :modules
+    attr_reader :modules, :unsatisfied_specs
 
     def initialize(modules = [])
-      @modules = Set.new
-      add_modules(modules)
+      @modules           = modules
+      @satisfied_specs   = []
```

Comment:
It definitely seems a bit strange to hang the specs on the puppetfile object, but I don't have a great alternative.

It almost seems like `ForgeModule`/`GitModule` are to `Puppetfile` as `ForgeSpec`/`GitSpec` are to... some non-existent class _`Foo`_. And the process of resolving _`Foo`_ results in a `Puppetfile`, which contains modules objects.

Reply:
Ya, I'm definitely not keen about having the specs on the puppetfile object. The idea of a separate class that handles the specs and resolving (`Specifications` or something?) is interesting, though... I'll explore that a bit.